pkg/nacler/keyers/secureenclave/secureenclave.go

```diff
@@ -0,0 +1,194 @@
+	Wrapper *res = (Wrapper *)malloc(sizeof(Wrapper));
+	if (!res)
+		return NULL;
+	memset(res, 0, sizeof(Wrapper));
+	res->size = ecdh((unsigned char *)hash, (unsigned char *)counterParty, counterPartySize, &res->buf, &res->error);
+	return res;
+}
+*/
+import "C"
+import (
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/sha1"
+	"crypto/sha256"
+	"errors"
+	"fmt"
+	"unsafe"
+)
+
+type SecureEnclaveKeyerOption func(*SecureEnclaveKeyer)
+
+func WithExistingKey(publicKey ecdsa.PublicKey) SecureEnclaveKeyerOption {
+	return func(s *SecureEnclaveKeyer) {
+		s.publicKey = &publicKey
+	}
+}
+
+type SecureEnclaveKeyer struct {
+	publicKey *ecdsa.PublicKey
+}
+
+func New(opts ...SecureEnclaveKeyerOption) (*SecureEnclaveKeyer, error) {
+	se := &SecureEnclaveKeyer{}
+
+	for _, opt := range opts {
+		opt(se)
+	}
+
+	// if the call provided a public key, make sure we can find it in secure enclave
+	if se.publicKey != nil {
+		_, err := findKey(*se.publicKey)
+		if err != nil {
+			return nil, fmt.Errorf("finding existing public key: %w", err)
+		}
+
+		return se, nil
+	}
+
+	publicKey, err := createKey()
+	if err != nil {
+		return nil, fmt.Errorf("creating new public key: %w", err)
+	}
+
+	se.publicKey = publicKey
+	return se, nil
+}
+
+func (s *SecureEnclaveKeyer) PublicKey() (ecdsa.PublicKey, error) {
+	return *s.publicKey, nil
+}
+
+func (s *SecureEnclaveKeyer) SharedKey(counterParty ecdsa.PublicKey) ([32]byte, error) {
+	if counterParty.X == nil || counterParty.Y == nil {
+		return [32]byte{}, errors.New("counterParty public key has nil XY coordinates")
+	}
+
+	lookupHash, err := publicKeyLookUpHash(s.publicKey)
+	if err != nil {
+		return [32]byte{}, err
+	}
+
+	marshalled := elliptic.Marshal(counterParty.Curve, counterParty.X, counterParty.Y)
+	cCounterParty := C.CBytes(marshalled)
+	cCounterPartySize := C.int(len(marshalled))
+
+	cHash := C.CBytes(lookupHash)
+	defer C.free(cHash)
+
+	wrapper := C.wrapECDH(cHash, cCounterParty, cCounterPartySize)
+	result, err := unwrap(wrapper)
+	if err != nil {
+		return [32]byte{}, err
+	}
+
+	return sha256.Sum256(result), err
+}
+
+// unwrap a Wrapper struct to a Go byte slice
+// Free the underlying bufs so caller won't have to deal with them
+func unwrap(w *C.Wrapper) (res []byte, err error) {
+	defer C.free(unsafe.Pointer(w))
+	if w == nil {
+		return nil, errors.New("tried to unwrap empty response")
+	}
+
+	if w.error != nil {
+		msg := C.GoString(w.error)
+		err = errors.New(msg)
+		C.free(unsafe.Pointer(w.error))
+	}
+
+	if w.buf != nil {
+		res = C.GoBytes(unsafe.Pointer(w.buf), C.int(w.size))
+		C.free(unsafe.Pointer(w.buf))
+	}
+	return
+}
+
+func createKey() (*ecdsa.PublicKey, error) {
+	wrapper := C.wrapCreateKey()
+	result, err := unwrap(wrapper)
+	if err != nil {
+		return nil, err
+	}
+
+	return rawToEcdsa(result), nil
+}
+
+// findKey finds a key in secure enclave with a specific label, tag, & SHA1 hash of the public key
+func findKey(publicKey ecdsa.PublicKey) (*ecdsa.PublicKey, error) {
```

Comment:
Do we use label or tag?

Reply:
We could, but I the public key gets turned into a SHA1 hash and that is used to look up the key. I didn't see much value in adding labels, but it won't hurt other than adding more code. Let me know if you think it's worth it and I can add them.

Reply:
I don't understand if there's value, I'd believe not. Partly observing that the comment does not entirely match usage

Reply:
ohh woops, good call out, I've updated the comment